[red-knot] Handle gradual intersection types in assignability

jgeralnik opened the pull request:
## Summary

This mostly fixes #14899

My motivation was similar to the last comment by @sharkdp there. I ran red_knot on a codebase and the most common error was patterns like this failing:

```
def foo(x: str): ...

x: Any = ...
if isinstance(x, str):
    foo(x) # Object of type `Any & str` cannot be assigned to parameter 1 (`x`) of function `foo`; expected type `str`
```

The desired behavior is pretty much to ignore Any/Unknown when resolving intersection assignability - `Any & str` should be assignable to `str`, and `str` should be assignable to `str & Any`
 
The fix is actually very similar to the existing code in `is_subtype_of`, we need to correctly handle intersections on either side, while being careful to handle dynamic types as desired.

This does not fix the second test case from that issue:

```
static_assert(is_assignable_to(Intersection[Unrelated, Any], Not[tuple[Unrelated, Any]]))
```

but that's misleading because the root cause there has nothing to do with gradual types. I added a simpler test case that also fails:

```
static_assert(is_assignable_to(Unrelated, Not[tuple[Unrelated]]))
```
This is because we don't determine that Unrelated does not subclass from tuple so we can't rule out this relation. If that logic is improved then this fix should also handle the case of the intersection

## Test Plan

Added a bunch of is_assignable_to tests, most of which failed before this fix.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/type_properties/is_assignable_to.md:
```diff
@@ -259,18 +260,36 @@ static_assert(is_assignable_to(Grandchild, Intersection[Child1, Child2]))
 static_assert(not is_assignable_to(int, Not[int]))
 static_assert(not is_assignable_to(int, Not[Literal[1]]))
 
 static_assert(not is_assignable_to(Intersection[Any, Parent], Unrelated))
```
This assertion is pre-existing, and wasn't labeled with a TODO, but it is wrong (which I'm sure made things confusing in terms of the desired behavior). `Any` can materialize to any type, which means that `Any & Parent` could be `Never & Parent` (that is, `Never`), or it could be `Unrelated & Parent`. Both of those are assignable to `Unrelated`. So `Any & Parent` should be assignable to `Unrelated`.

jgeralnik replied:
I'm not 100% sure I agree. Take the following code:

```
def f(x: str): ...

def g(x: Any):
    if isinstance(x, str):
         f(x)

def h(x: Any):
    if isinstance(x, int):
        f(x)
```

The goal of this PR is to solve the call in g, which is done because we allow assigning str & Any to str.

Should the call in h also succeed or should we raise an error?

I can see the argument you're making - just because x is an int doesn't mean it's not a str, so in theory a caller could pass an str & int into h, and since h is untyped we can't assume what arguments it gets.

On the other hand, the more helpful thing to the enduser is probably to effectively assume that x is an int here and not allow the assignment, since it's almost certainly incorrect.

For what it's worth mypy seems to agree with me, and raises an error here:

```
error: Argument 1 to "f" has incompatible type "int"; expected "str"  [arg-type]
```

carljm replied:
Technically, the call in `h` should fail, once we add the ability (which we don't yet have) to recognize that certain C extension or builtin types (or Python types with `__slots__`) can't be multiply-inherited together at runtime due to incompatible memory layout. 

But with two arbitrary types `A` and `B` (which are neither final nor use `__slots__`) in place of `int` and `str`, the call in `h` should succeed. That's simply the nature of `Any`; there's no consistent theoretical basis for any other behavior. Not implementing this assignability breaks the gradual guarantee (which is a foundational -- perhaps the foundational -- principle of gradual typing): that replacing a static type annotation with `Any` should not cause new type errors to appear. It's easy to see that in the following code the gradual guarantee would be broken:

```py
class A: pass
class B: pass

def f(x: B):
    pass

def h(x: B):
    if isinstance(x, A):
        f(x)
```

This code should type-check as-is, because `A & B` is assignable to `B`. If we don't consistently implement `Any` in intersections, replacing the annotation `x: B` with `x: Any` would cause a new type error to appear. This violates the gradual guarantee.

Mypy does not really implement intersection types, its narrowing behavior is more ad-hoc, and it has decided to incorrectly eliminate `Any` in intersections, but we are not going to follow that behavior.

jgeralnik replied:
Alright, makes sense as a design philosophy. The alternative philosophy is "provide as much value as reasonably possible to a codebase without typing" but I can understand why it makes sense to optimize on soundness.

Will update the PR

carljm commented on crates/red_knot_python_semantic/resources/mdtest/type_properties/is_assignable_to.md:
```diff
@@ -259,18 +260,36 @@ static_assert(is_assignable_to(Grandchild, Intersection[Child1, Child2]))
+static_assert(is_assignable_to(Intersection[Any, Parent], Parent))
+static_assert(is_assignable_to(Intersection[Any, Child1], Parent))
+static_assert(is_assignable_to(Intersection[Any, Child2, Not[Child1]], Parent))
+static_assert(not is_assignable_to(Intersection[Any, Parent], Unrelated))
```
This is a duplicate of the line commented above, and is also wrong; this should be assignable.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/type_properties/is_assignable_to.md:
```diff
@@ -259,18 +260,36 @@ static_assert(is_assignable_to(Grandchild, Intersection[Child1, Child2]))
-# TODO: The following assertions should not fail (see https://github.com/astral-sh/ruff/issues/14899)
-# error: [static-assert-error]
-static_assert(is_assignable_to(Intersection[Any, int], int))
+# Intersection with `Any` is effectively ignored for the sake of assignment
```
This comment isn't right. Intersection with `Any` is not ignored, rather it dominates, in the sense that `Any & T` is assignable to any type, regardless of `T`. (`Any & T` is even assignable to a type `S` that is disjoint from `T`, because `Any` could represent `S`, and `S & T` is `Never`, which is assignable to `S`.)
```suggestion
```

carljm commented on crates/red_knot_python_semantic/resources/mdtest/type_properties/is_assignable_to.md:
```diff
@@ -259,18 +260,36 @@ static_assert(is_assignable_to(Grandchild, Intersection[Child1, Child2]))
+static_assert(is_assignable_to(Parent, Parent | Intersection[Any, Unrelated]))
+
+static_assert(not is_assignable_to(Intersection[Child1, Child2], Intersection[Parent, Unrelated]))
+static_assert(not is_assignable_to(Intersection[Parent, Any], Intersection[Parent, Unrelated]))
```
Similar to the above cases, this should be assignable.

carljm commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -802,6 +802,29 @@ impl<'db> Type<'db> {
+            }
+
+            // An intersection type S is assignable to a type T if
+            // Any non-dynamic element of S is assignable to T (e.g. `int & Any` is assignable to `int` (but not to `str`))
+            // Negative elements do not have an effect on assignability - if S is assignable to T then S & ~P is also assignable to T.
+            (Type::Intersection(intersection), ty) => {
+                intersection.positive(db).iter().any(|&elem_ty| {
+                    !matches!(elem_ty, Type::Dynamic(_)) && elem_ty.is_assignable_to(db, ty)
+                })
+            }
```
As discussed above, the special handling of dynamic types here is not correct. I think the right behavior falls out from just simplifying this code to eliminate that special case.
```suggestion
            // Any element of S is assignable to T (e.g. `int & Any` is assignable to `int` (but not to `str`))
            // Negative elements do not have an effect on assignability - if S is assignable to T then S & ~P is also assignable to T.
            (Type::Intersection(intersection), ty) => {
                intersection.positive(db).iter().any(|&elem_ty| {
                    elem_ty.is_assignable_to(db, ty)
                })
            }
```

carljm reviewed (commented):
Nice work, and thank you for the contribution!

This mostly looks right, I think there's just one area that needs tweaking.

jgeralnik commented:
Fixed the code, which also fixed a diagnostic we were raising in ruff_benchmark. This fixed diagnostic completely convinced me that you are right about the desired behavior, by the way - the code was effectively:

```
def f(x: str): ...

x: Any
if x:
    reveal_type(x) # Unknown & ~AlwaysFalsy | @Todo & ~AlwaysFalsy
    f(x)
```

which my original version incorrectly typed as effectively equivalent to ~AlwaysFalsy and failed on because ~AlwaysFalsy is not assignable to str.

carljm reviewed (approved):
Awesome, thank you so much for the PR!
